crabs/detection_tracking/detection_utils.py

```diff
@@ -226,65 +229,243 @@ def calculate_iou(box1, box2) -> float:
-        gt_annotations = gt_data.get("annotations", [])
-        for gt_box in gt_annotations:
-            if gt_box["image_id"] == frame_number:
-                x_gt, y_gt, width_gt, height_gt = gt_box["bbox"]
-                x2_gt, y2_gt = x_gt + width_gt, y_gt + height_gt
-                x_gt, y_gt, x2_gt, y2_gt = (
-                    int(x_gt),
-                    int(y_gt),
-                    int(x2_gt),
-                    int(y2_gt),
-                )
+    # Bounding box geometry
+    xmin, ymin, xmax, ymax, id = bbox
+    width_box = int(xmax - xmin)
+    height_box = int(ymax - ymin)
+
+    # Add to csv
+    csv_writer.writerow(
+        (
+            frame_name,
+            frame.size,
+            '{{"clip":{}}}'.format("123"),
+            1,
+            0,
+            '{{"name":"rect","x":{},"y":{},"width":{},"height":{}}}'.format(
+                xmin, ymin, width_box, height_box
+            ),
+            '{{"track":"{}"}}'.format(id),
+        )
+    )
 
-                for bbox in tracked_boxes:
-                    x1_track, y1_track, x2_track, y2_track, _ = bbox
-                    iou = calculate_iou(
-                        [x_gt, y_gt, x2_gt, y2_gt],
-                        [x1_track, y1_track, x2_track, y2_track],
-                    )
 
-                    if iou > iou_threshold:
-                        cv2.rectangle(
-                            frame_copy,
-                            (x_gt, y_gt),
-                            (x2_gt, y2_gt),
-                            (0, 255, 0),
-                            2,
-                        )
-                        id_label = f"id : {object_id}"
-                        draw_bbox(
-                            frame_copy,
-                            int(x1_track),
-                            int(y1_track),
-                            int(x2_track),
-                            int(y2_track),
-                            (0, 0, 255),
-                            id_label,
-                        )
+def save_frame_and_csv(
+    video_file_root: str,
+    tracking_output_dir: Path,
+    tracked_boxes: List[List[float]],
+    frame: np.ndarray,
+    frame_number: int,
+    csv_writer: Any,
+    save_plot: bool = True,
+) -> np.ndarray:
+    """
+    Common functionality for saving frames and CSV
+    """
+    frame_copy = frame.copy()
+
+    for bbox in tracked_boxes:
+        # Get frame name
+        frame_name = f"{video_file_root}frame_{frame_number:08d}.png"
+
+        # Add bbox to csv
+        write_bbox_to_csv(bbox, frame, frame_name, csv_writer)
+
+        # Save frame as PNG
+        frame_path = tracking_output_dir / frame_name
+        img_saved = cv2.imwrite(str(frame_path), frame)
+        if img_saved:
+            logging.info(f"Frame {frame_number} saved at {frame_path}")
+        else:
+            logging.info(
+                f"ERROR saving {frame_name}, frame {frame_number}...skipping"
+            )
+            break
```

Comment:
```suggestion
        if not img_saved:
            logging.error(f"Didn't save {frame_name}, frame {frame_number}, Skipping.")
            break
        logging.info(f"Frame {frame_number} saved at {frame_path}")
```
Or is it only a [logger.warning](https://docs.python.org/3/library/logging.html#logging.WARNING), since we continue processing?

----

Also: guarding if statements are arguably better style.

```
if bad_thing:
    logger.error("bad thing")

if terrible_thing:
    raise Exception("aaaaah")

do_stuff()
logger.info("everything is fine")
```
You can add more guards as more edge cases appear without complicating a huge if statement.
```
if (not bad_thing or not terrible_thing ...) # this line becomes unreadable
    do_stuff()
elif bad_thing:
    ....
elif terrible_thing:
    ...
else:
    ...
 ```